doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,232 @@
+storage pick the timestamp. Having storage select the timestamp aligns better with our goal of
+having a storage owned "push source" in the future.
+
+### Request Limits
+[request_limits]: #request_limits
+
+To start we'll aim to support at least 100 requests per second, in other words, 10 sources could
+receive 10 requests per second. We can probably scale further, but this seems like a sensible limit
+to start at. Also we'll add an [`axum::middleware`](https://docs.rs/axum/latest/axum/middleware/index.html)
+that enforces rate limiting on our new endpoint. This should be a pretty easy way to prevent a
+misbehaving source from taking down all of `environmentd`.
+
+### Metrics
+[metrics]: #metrics
+
+There are three metrics we should track, to measure success and performance of this new source:
+
+1. Number of `HTTP REQUESTS` sources created.
+2. Number of requests made to the `/api/webhook` endpoint.
+3. Response time of the `/api/webhook` endpoint.
+
+A second way to measure success, and get feedback on this new feature, is to use it internally.
+To start, we plan to dogfood the new source with the following applications:
+
+1. Segment
+2. GitHub
+3. Slack
+4. Buildkite
+
+This internal usage will help us quickly determine if we built the right thing, and how to
+prioritize further improvements.
+
+## Alternatives
+
+### Fivetran
+
+For the idea of "making it easier to get data into Materialize", an alternative approach we
+considered was building Fivetran support. While this is feasible, and something we might pursue in
+the future, we're not pursuing this path for two reasons:
+
+1. It requires more work to build.
+2. The at best 5 minute sync time of Fivetran, does not align well with being an "operational" data
+warehouse.
+
+### `environmentd` vs `clusterd`
+
+In the proposed design, `environmentd` handles all of the incoming requests, which is not good for
+scalability. An alternative is to have a `clusterd` handle requests for a given Source, like every
+other Source does today. The reason we chose `environmentd` to primarily handle webhook requests is
+because it is already setup to handle _incoming_ requests from an external network. All of our
+other sources are pull based, where we send a request and ask for data. Whereas webhooks are push
+based, we receive a request, when there is new data available. Exposing a `clusterd` outside of our
+network, maybe isn't even something we want to do (e.g. in the future maybe there is a reverse
+proxy that talks to `clusterd`), and would require significant lift from the Cloud Team.
+
+### Tables vs Sources vs What we Chose
+
+An alternative approach is to use a `TABLE` for these new sources, and "translate" any incoming
+requests to `INSERT` statements. There are two reasons we did not chose this approach:
+
+1. `INSERT`s into a table are linearizable, whereas requests from a webhook don't need to be. By
+using tables we would be imposing an unnecessary performance bottleneck.
```

Comment:
I don't think I agree with this point. Table's don't need to have linearizable writes either, we've just chosen to make writes linearizable at the cost of performance. If this tradeoff imposes unnecessary performance bottlenecks, then that is more of an argument for tables writes to not be linearizable (or have the linearizability be configurable) rather than invent a new object type to circumvent linearizable writes.

I think the next point is a much more convincing reason to not use tables.

Reply:
Good point! I updated the doc to emphasize the other point more